gen/qr.hoon

```diff
@@ -0,0 +1,359 @@
+::
+::  TODO:  implement the other 7 masks
+::         penalty scoring
+::         move the "two '█' per module" logic up to pretty-print
+::
+::  TOC::...
+::       :1: main generator
+::       :2: types
+::       :3: main logic
+::       :4: qr formatting helpers
+::       :5: message encoding helpers
+::       :6: long constants
+::
+::  implemented with much help from the following sites:
+::    https://urbit.org/docs/
+::    https://www.thonky.com/qr-code-tutorial/
+::    https://en.wikiversity.org/wiki/Reed%E2%80%93Solomon_codes_for_coders
+::
+::::  /~somrem-bosdes/home/gen/qr/hoon  ::::::::::::::::::::::::::::::::::::::::
+::
+::::::::::::::::::::::::::::::::::::::::::::::::1: main generator
+::
+:-  %say
+|=  [* [msg=tape ~] ~ ~]
+::
+?:  (gth (lent msg) 106)
+  :-  %tape
+      "Tape must be 106 characters or less, yours is {<(lent msg)>} characters"
+=-
+:-  %tang
+  %-  pretty-print.qr
+  %-  encode.qr  msg
+::
+^=  qr  |%
+::
+::::::::::::::::::::::::::::::::::::::::::::::::2: types
+::
+::  TODO type this better, will ease implementation of new features
+::
+++  grid  (list (list ?))
+::
+::::::::::::::::::::::::::::::::::::::::::::::::3: main logic
+::
+++  pretty-print
+  |=  qr=grid
+  ^-  tang
+  =/  rq  (flop qr)  :: to get around `tang` printing weirdness
+  %+  turn  rq
+  |=  row/(list ?)
+  :-  %leaf
+  ^-  tape
+  %+  turn  row
+  |=  solid/?
+  ?:  solid
+    ' '
+  '█'
+::
+++  encode
+  |=  message=tape
+  ^-  grid
+  %-  add-border
+  %-  add-mask-information
+  %-  pick-mask
+  %-  add-data  
+  :-  %-  add-error-correction 
+          message
+  %-  create-base
+    "5L"
+::
+::::::::::::::::::::::::::::::::::::::::::::::::4: qr formatting helpers
+::
+++  create-base
+  |=  version=tape
+  ^-  grid
+  ?.  =(version "5L")
+    ~&  [%fatal-error "sorry, only QR version 5L is currently supported"]
+    !!
+  %-  atm  
+  :-  ;:  weld 
+    dark-module
+    timing-pattern 
+    alignment-pattern
+    finder-pattern
+  ==
+::
+  %-  init-qr-base
+  37  :: hardcoding QR version 5-L
+::
+++  init-qr-base
+  |=  n=@u
+  ^-  grid
+  %+  reap  n     
+  %+  reap  (mul 2 n)  :: one '█' is too narrow, use two together for each "module" 
+                       :: TODO yikes module abstraction needed. let pretty-print do this?
+      %.n
+::
+++  add-border
+  |=  qr=grid
+  ^-  grid
+  =/  n=@u  (lent qr)
+  =/  i=@u  0
+  :-  (reap (mul 2 (add 2 n)) %.n)
+  |-
+  ?:  =(i n)
+    :-  (reap (mul 2 (add 2 n)) %.n)  ~
+  :-  (zing [[%.n %.n ~] (snag i qr) [%.n %.n ~] ~])
+  $(i (add 1 i))
+::
+++  add-mask-information
+  |=  [mask=@u qr=grid]
+  ^-  grid
+  =/  info  (format-strings mask)
+  =/  i=@u  0  :: TODO dont hardcode mask 000
+  |-
+  ?:  =(i (lent info))
+    qr
+  ?:  =((snag i info) 1)
+    =/  coords-one  (snag i info-route-one)
+    =/  coords-two  (snag i info-route-two)
+    %-  tm  :-  [-.coords-one +.coords-one]
+    %-  tm  :-  [-.coords-two +.coords-two]
+    $(i (add 1 i))
```

Comment:
Since `tm` takes two arguments, it's clearer to use `%+` instead. (That also removes the need for explicit `:-`.)  
Also, `[-.a +.a]`, as in `[head-of-a, tail-of-a]`, is identical to just `a`. So;

```hoon
%+  tm  coords-two
$(i (add 1 i))
```